### PR TITLE
Loading Pull Request List via GraphQL

### PR DESCRIPTION
**Pull request by StanleyGoldman**
In order to implement (#1020), this pull request changes functionality to query for the pull request list via GraphQL. REST API calls to obtain Pull Request list items do not include the `mergeable` flag.

### CODE REVIEW

File: src/view/treeNodes/categoryNode.ts [hunk 1]
```diff
@@ -73,17 +73,11 @@ export class PRCategoryActionNode extends TreeNode implements vscode.TreeItem {
 	}
 }
 
-interface PageInformation {
-	pullRequestPage: number;
-	hasMorePages: boolean;
-}
```
Comment (StanleyGoldman) [inline src/view/treeNodes/categoryNode.ts]: This seems to be an artifact from an old implementation.

File: src/view/treeNodes/categoryNode.ts [hunk 2]
```diff
@@ -73,17 +73,11 @@ export class PRCategoryActionNode extends TreeNode implements vscode.TreeItem {
...
-
 export class CategoryTreeNode extends TreeNode implements vscode.TreeItem {
 	public readonly label: string;
 	public collapsibleState: vscode.TreeItemCollapsibleState;
 	public prs: PullRequestModel[];
 	public fetchNextPage: boolean = false;
-	public repositoryPageInformation: Map<string, PageInformation> = new Map<string, PageInformation>();
```
Comment (StanleyGoldman) [inline src/view/treeNodes/categoryNode.ts]: As does this.

File: src/github/pullRequestManager.ts [hunk 3]
```diff
@@ -582,15 +582,81 @@ export class PullRequestManager {
...
+						pullRequestFilter = pr => pr.reviewRequests.nodes
+							.findIndex(reviewRequest => reviewRequest.requestedReviewer.login === currentUserLogin) !== -1;
+					} else if (type === PRType.AssignedToMe) {
+						pullRequestFilter = pr => pr.assignees.nodes
+							.findIndex(asignee => asignee.login === currentUserLogin) !== -1;
+					} else {
+						throw new Error('Unexpected pull request filter');
+					}
+
+					pullRequestItems = pullRequestItems.filter(pullRequestFilter);
```
Comment (StanleyGoldman) [inline src/github/pullRequestManager.ts]: Client side filtering happening here.

File: src/github/pullRequestManager.ts [hunk 4]
```diff
@@ -582,15 +582,81 @@ export class PullRequestManager {
...
+							body: '',
+							createdAt: pullRequestItem.createdAt,
+							updatedAt: pullRequestItem.updatedAt,
+							merged: pullRequestItem.merged,
+							nodeId: pullRequestItem.nodeId,
+							number: pullRequestItem.number,
+							state: pullRequestItem.state,
+							title: pullRequestItem.title,
+							user: pullRequestItem.author
+						};
```
Comment (StanleyGoldman) [inline src/github/pullRequestManager.ts]: Mapping the graphql object to an object the `PullRequestModel` consumes. Not thrilled about doing this here.

File: src/github/graphql.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import { IAccount } from "./interface";
+
```
Comment (rebornix) [inline src/github/graphql.ts]: I guess it's from auto import and it misplaces the code ;)

File: src/github/graphql.ts [hunk 6]
```diff
@@ -309,4 +311,66 @@ export interface RateLimit {
...
+	};
+	reviewRequests: {
+		nodes: [{
+			requestedReviewer: IAccount;
+		}];
+	};
+	createdAt: string;
+	updatedAt: string;
+	merged: boolean;
+	headRef: {
```
Comment (rebornix) [inline src/github/graphql.ts]: Can we somehow reuse `interface Ref`

Comment (rebornix) [reply src/github/graphql.ts]: Or probably a new `interface RefParts`

File: src/github/graphql.ts (hunk 5 above)
Comment (StanleyGoldman) [reply src/github/graphql.ts]: Thanks, fixed.

File: src/github/graphql.ts (hunk 6 above)
Comment (StanleyGoldman) [reply src/github/graphql.ts]: Yea, it's good enough to re-use `Ref` thanks.

File: src/github/githubRepository.ts [hunk 7]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
 		};
 	}
 
-	async getPullRequests(prType: PRType, page?: number): Promise<PullRequestData | undefined> {
-		return prType === PRType.All ? this.getAllPullRequests(page) : this.getPullRequestsForCategory(prType, page);
-	}
+	async getPullRequestsGraphQL(type: PRType, nextCursor?: string|null):Promise<PullRequestListResponse|undefined> {
+		const { remote, query, octokit } = await this.ensure();
+		const currentUser = octokit && (octokit as any).currentUser;
```
Comment (annthurium) [inline src/github/githubRepository.ts]: just out of curiosity, why `octokit as any`?  Do we not have ts bindings for octokit or something?

File: src/github/githubRepository.ts [hunk 8]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
...
-							);
-							return null;
-						}
-
-						const item = convertRESTPullRequestToRawPullRequest(pullRequest);
-						return new PullRequestModel(this, this.remote, item);
-					}
-				)
-				.filter(item => item !== null) as PullRequestModel[];
+		let filter = `type:pr is:open repo:${remote.owner}/${remote.repositoryName}`;
```
Comment (annthurium) [inline src/github/githubRepository.ts]: Might be a little edge casey, but do we need to guard against the remote being `null` here?

File: src/github/githubRepository.ts [hunk 9]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
...
+		if (type !== PRType.All) {
+			if (type === PRType.Mine) {
+				filter += ` author:${currentUserLogin}`;
+			} else if (type === PRType.RequestReview) {
+				filter += ` review-requested:${currentUserLogin}`;
+			} else if (type === PRType.AssignedToMe) {
+				filter += ` assignee:${currentUserLogin}`;
 			} else {
-				throw e;
+				throw new Error('Unexpected pull request filter');
```
Comment (annthurium) [inline src/github/githubRepository.ts]: would it make sense to log the filter type here?

File: src/github/githubRepository.ts (hunk 7 above)
Comment (StanleyGoldman) [reply src/github/githubRepository.ts]: I have a similar question as I'm pretty sure the bindings exist.

Comment (RMacfarlane) [reply src/github/githubRepository.ts]: Yeah, we do have bindings for octokit, but `currentUser` is something we've added ourselves

https://github.com/Microsoft/vscode-pull-request-github/pull/600#pullrequestreview-167227402

We also added an `octokit.ts` file a while ago that modifies the typings slightly, there was something that should have been nullable that wasn't. `currentUser` can be added there to get rid of our `any` casts. But I think we should ultimately try to get rid of that file by correcting the typings upstream

File: src/github/githubRepository.ts [hunk 10]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
...
 
-	async getPullRequests(prType: PRType, page?: number): Promise<PullRequestData | undefined> {
-		return prType === PRType.All ? this.getAllPullRequests(page) : this.getPullRequestsForCategory(prType, page);
-	}
+	async getPullRequestsGraphQL(type: PRType, nextCursor?: string|null):Promise<PullRequestListResponse|undefined> {
+		const { remote, query, octokit } = await this.ensure();
+		const currentUser = octokit && (octokit as any).currentUser;
+		const currentUserLogin: string = currentUser.login;
 
-	private async getAllPullRequests(page?: number): Promise<PullRequestData | undefined> {
```
Comment (RMacfarlane) [inline src/github/githubRepository.ts]: I don't think we should remove this - we still need to support GHE that doesn't have GraphQL

File: src/github/pullRequestManager.ts (hunk 4 above)
Comment (RMacfarlane) [reply src/github/pullRequestManager.ts]: the file `src/github/utils.ts` has functions for mapping response types to normalized types

File: src/github/githubRepository.ts (hunk 8 above)
Comment (StanleyGoldman) [reply src/github/githubRepository.ts]: I would say yes, but looking at all the other calls to `ensure` to get `remote` seem to make the assumption that remote will not be null.

File: src/github/githubRepository.ts (hunk 9 above)
Comment (StanleyGoldman) [reply src/github/githubRepository.ts]: :+1:

File: src/github/pullRequestManager.ts (hunk 4 above)
Comment (StanleyGoldman) [reply src/github/pullRequestManager.ts]: Thanks, i'll move the code.

File: src/github/githubRepository.ts [hunk 11]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
...
-			Logger.appendLine(`Fetching all pull requests failed: ${e}`, GitHubRepository.ID);
-			if (e.code === 404) {
-				// not found
-				vscode.window.showWarningMessage(`Fetching pull requests for remote '${this.remote.remoteName}' failed, please check if the url ${this.remote.url} is valid.`);
+		if (type !== PRType.All) {
+			if (type === PRType.Mine) {
+				filter += ` author:${currentUserLogin}`;
+			} else if (type === PRType.RequestReview) {
+				filter += ` review-requested:${currentUserLogin}`;
+			} else if (type === PRType.AssignedToMe) {
```
Comment (StanleyGoldman) [inline src/github/githubRepository.ts]: asdf

File: src/github/githubRepository.ts [hunk 12]
```diff
@@ -203,103 +203,44 @@ export class GitHubRepository implements IGitHubRepository, vscode.Disposable {
...
-			};
-		} catch (e) {
-			Logger.appendLine(`GitHubRepository> Fetching all pull requests failed: ${e}`);
-			if (e.code === 404) {
-				// not found
-				vscode.window.showWarningMessage(`Fetching pull requests for remote ${this.remote.remoteName}, please check if the url ${this.remote.url} is valid.`);
-			} else {
-				throw e;
-			}
+		if(!!nextCursor) {
```
Comment (smbkr) [inline src/github/githubRepository.ts]: Just curious why you use `if(!!condition)` here, instead of `if(condition)`

**Comment by alexr00** [comment]
Thanks for the PR @StanleyGoldman. Since the goal of this PR was to enable https://github.com/microsoft/vscode-pull-request-github/issues/681, I'm going to close it. We already have https://github.com/microsoft/vscode-pull-request-github/issues/681 as of the last stable release.

**Pull request closed without merging**